### PR TITLE
Contour fixes/improvements

### LINKED ISSUE #16153: gap size for contour labels is poorly estimated
### Bug report

**Bug summary**

The size of the gap for contour labels in contour plots is poorly estimated.  (Noted while looking at #8819).

**Code for reproduction**

This uses a custom "very wide" font to make the problem more visible: download the font at https://www.1001fonts.com/20-db-font.html, put it somewhere where font_manager will find it, and run
```python
from pylab import *
# You'll need to regen your font cache first.
rcParams['font.family'] = '20 db'
plt.figure()
x = np.arange(400).reshape((20,20))
c = plt.contour(x)
plt.clabel(c)
plt.show()
```

**Actual outcome**

bad

**Expected outcome**

good

---

The fix is actually not that hard (that's how I gen'd the "good" version)
```diff
diff --git a/lib/matplotlib/contour.py b/lib/matplotlib/contour.py
index bc17765c5..dd07d9f22 100644
--- a/lib/matplotlib/contour.py
+++ b/lib/matplotlib/contour.py
@@ -236,20 +236,12 @@ class ContourLabeler:
         """
         if not isinstance(lev, str):
             lev = self.get_text(lev, fmt)
-        lev, ismath = text.Text()._preprocess_math(lev)
-        if ismath == 'TeX':
-            lw, _, _ = (texmanager.TexManager()
-                        .get_text_width_height_descent(lev, fsize))
-        elif ismath:
-            if not hasattr(self, '_mathtext_parser'):
-                self._mathtext_parser = mathtext.MathTextParser('bitmap')
-            img, _ = self._mathtext_parser.parse(lev, dpi=72,
-                                                 prop=self.labelFontProps)
-            _, lw = np.shape(img)  # at dpi=72, the units are PostScript points
-        else:
-            # width is much less than "font size"
-            lw = len(lev) * fsize * 0.6
-        return lw
+        fig = self.ax.figure
+        width = (text.Text(0, 0, lev,
+                           figure=fig, fontproperties=self.labelFontProps)
+                 .get_window_extent(fig.canvas.get_renderer()).width)
+        width *= 72 / fig.dpi
+        return width
 
     def set_label_props(self, label, text, color):
         """Set the label properties - color, fontsize, text."""
```
but this breaks a bunch of baseline images (unsurprisingly) so I'm going to combine this with other baseline-breaking cleanups I want to do on contour... Specifically, I plan to
- get rid of https://github.com/matplotlib/matplotlib/blob/262c67fad54608e1a8ee2b7c2aac8e9e5f01dba2/lib/matplotlib/contour.py#L564-L567 (which can be removed with no bad effect)
- check whether we can make the default formatter of labels the same formatter as the one that would be used for a colorbar; this would allow the use of a unicode minus and other benefits listed in https://github.com/matplotlib/matplotlib/pull/12459 (see also https://github.com/matplotlib/matplotlib/pull/12473).

---

**Matplotlib version**

  * Operating system: linux
  * Matplotlib version: master
  * Matplotlib backend (`print(matplotlib.get_backend())`): agg
  * Python version: 38
  * Jupyter version (if applicable):
  * Other libraries:

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

Closes #16153, and does some more.

Three fixes/improvements to contour labeling, grouped together to minimize baseline images changes.  ~~Code-wise I think this is ready for review, but marking as draft PR just so that I can still have a look at whether we can combine/delete some of the baseline images.~~ I also took advantage of the baseline regen to merge two tests (test_contour_manual_labels and test_contour_label_size_color).

Commit 1: Improve estimation of label text width, as explained in #16153.

Commit 2: Remove https://github.com/matplotlib/matplotlib/blob/262c67fad54608e1a8ee2b7c2aac8e9e5f01dba2/lib/matplotlib/contour.py#L564-L567 which shifts the labels a bit but has no bad effects (there is no division in print_label and division by zero warning is expressly suppressed in locate_label.

Commit 3: Switch the default formatter to ScalarFormatter and use format_ticks to format all ticks at once, allowing the reuse of auto-precision choice by ScalarFormatter.  Ultimately we could even use someting like #12473 (reusing colorbar logic to format values) but this shouldn't change any baseline images as none of them uses log scale (so they would still all use ScalarFormatter).

Commit 4: Update baseline images all at once.  See the precision change and unicode minus sign in e.g. test_contour/contour_test_label_transforms.png and the tighter fit of labels in e.g. test_patheffects/collection.png.

~~I'm not sure why CI is failing even though I updated the baseline images locally, though.~~ Figured out the test failure.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/contour.py [hunk 1]
```diff
@@ -92,14 +90,16 @@ def clabel(self, levels=None, *,
...
+            How the levels are formatted:
 
-            Alternatively, this can be a dictionary matching contour levels
-            with arbitrary strings to use for each contour level (i.e.,
-            fmt[level]=string), or it can be any callable, such as a
-            `.Formatter` instance, that returns a string when called with a
-            numeric contour level.
+            - If a `.Formatter`, it is used to format all levels at once, using
+              its `.Formatter.format_ticks` method.  The default is to use a
+              standard `.ScalarFormatter`.
```
Comment (QuLogic) [inline lib/matplotlib/contour.py:98]: I would put this default at the end outside of the list; it is not a default specific to this list item.

File: doc/users/next_whats_new/2020-01-09-AL.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Contour plots now default to using ScalarFormatter
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Pass ``fmt="%1.3f"`` to the countouring call to restore the old default label
```
[the remark this answers is not in the document]
Comment (anntzer) [reply doc/users/next_whats_new/2020-01-09-AL.rst:4]: yes

File: lib/matplotlib/contour.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/contour.py:98]: yes

File: doc/users/next_whats_new/2020-01-09-AL.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (QuLogic) [reply doc/users/next_whats_new/2020-01-09-AL.rst]: It says coutouring now...

[the remark this answers is not in the document]
Comment (anntzer) [reply doc/users/next_whats_new/2020-01-09-AL.rst]: second try...

**Review by QuLogic — APPROVED** [review]

**Pull request merged**